window/src/os/xdg_desktop_portal.rs

```diff
@@ -48,8 +61,20 @@ fn value_to_appearance(value: OwnedValue) -> anyhow::Result<Appearance> {
 }
 
 pub async fn get_appearance() -> anyhow::Result<Appearance> {
-    let value = read_setting("org.freedesktop.appearance", "color-scheme").await?;
-    value_to_appearance(value)
+    let mut state = STATE.lock().unwrap();
+    match state.appearance {
+        Some(a) => Ok(a),
+        None => {
+            let value = read_setting("org.freedesktop.appearance", "color-scheme").await?;
+            match value_to_appearance(value) {
+                Ok(a) => {
+                    state.appearance = Some(a);
+                    Ok(a)
+                }
+                err => err,
+            }
+        }
+    }
 }
 
 pub fn subscribe() {
```

Comment:
there's a race condition in this implementation: if the appearance is changed between our first call to `get_appearance` and before we have set up a subscription, we will not be aware of the change because `get_appearance` is not aware of whether the subscribe function is running and will blindly continue to use its cached value.

To resolve this race I would suggest adding a call to `read_setting` immediately before the `while` loop below (in between the `receive_SettingChanged` and the `while` loop).

If the returned value is different from the cache value, then `conn.advise_of_appearance_change(appearance);` should be called to ensure that the updated value is observed.

Reply:
I will implement your other suggested change later today, but I see a problem with the above idea.
If we call `read_setting` again before the while loop (in addition to the `read_setting` in `get_apperance`), then we have at least one duplicated dbus call again. Now in this case it wouldn't actually block startup because it would run in a separate thread, but it still sounds rather backwards to me since we could easily solve this in another way.

For example why can't we just solve the race condition by subscribing before calling the first `get_appearance`?

Reply:
get_appearance is often called as part of loading the config which happens before the gui subsystem is started up; subscribe cannot be called until after the config is processed at least once.